src/python/pants/backend/javascript/goals/test.py

```diff
@@ -0,0 +1,245 @@
+from pants.util.logging import LogLevel
+
+
+@dataclass(frozen=True)
+class JSCoverageData(CoverageData):
+    snapshot: Snapshot
+    address: Address
+    output_files: tuple[str, ...]
+    output_directories: tuple[str, ...]
+    working_directory: str
+
+
+class JSCoverageDataCollection(CoverageDataCollection[JSCoverageData]):
+    element_type = JSCoverageData
+
+
+@dataclass(frozen=True)
+class JSTestFieldSet(TestFieldSet):
+    required_fields = (JSTestSourceField,)
+
+    source: JSTestSourceField
+    dependencies: Dependencies
+    timeout: JSTestTimeoutField
+    extra_env_vars: JSTestExtraEnvVarsField
+
+
+class JSTestRequest(TestRequest):
+    tool_subsystem = NodeJSTest
+    field_set_type = JSTestFieldSet
+
+
+@rule(level=LogLevel.DEBUG, desc="Run javascript tests")
+async def run_javascript_tests(
+    batch: JSTestRequest.Batch[JSTestFieldSet, Any],
+    test: TestSubsystem,
+    test_extra_env: TestExtraEnv,
+) -> TestResult:
+    field_set = batch.single_element
+    installation_get = Get(InstalledNodePackage, InstalledNodePackageRequest(field_set.address))
+    transitive_tgts_get = Get(TransitiveTargets, TransitiveTargetsRequest([field_set.address]))
+
+    field_set_source_files_get = Get(SourceFiles, SourceFilesRequest([field_set.source]))
+    target_env_vars_get = Get(
+        EnvironmentVars, EnvironmentVarsRequest(field_set.extra_env_vars.sorted())
+    )
+    installation, transitive_tgts, field_set_source_files, target_env_vars = await MultiGet(
+        installation_get, transitive_tgts_get, field_set_source_files_get, target_env_vars_get
+    )
+
+    sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            (tgt.get(SourcesField) for tgt in transitive_tgts.closure),
+            enable_codegen=True,
+            for_sources_types=[JSSourceField, AssetSourceField],
+        ),
+    )
+    merged_digest = await Get(Digest, MergeDigests([sources.snapshot.digest, installation.digest]))
+
+    def relative_package_dir(file: str) -> str:
+        return os.path.relpath(file, installation.project_env.package_dir())
+
+    test_script = installation.project_env.ensure_target()[NodePackageTestScriptField].value
+    entry_point = test_script.entry_point
+
+    coverage_args: tuple[str, ...] = ()
+    output_files: list[str] = []
+    output_directories: list[str] = []
+    if test.use_coverage and test_script.supports_coverage():
+        coverage_args = test_script.coverage_args
+        output_files.extend(test_script.coverage_output_files)
+        output_directories.extend(test_script.coverage_output_directories)
+        entry_point = test_script.coverage_entry_point or entry_point
+
+    process = await Get(
+        Process,
+        NodeJsProjectEnvironmentProcess(
+            installation.project_env,
+            args=(
+                "run",
+                entry_point,
+                "--",
+                *sorted(map(relative_package_dir, field_set_source_files.files)),
```

Comment:
This feels like a minor abuse of `map`... would [use a generator instead](https://www.pantsbuild.org/docs/style-guide#prefer-comprehensions).